### DISCUSSION TITLE
Glob including variables

### DISCUSSION THREAD

**Discussion by cronhan** [nushell/nushell#12798] [Q&A] [answered]
Hi,
I think there is something fundamental that has not yet clicked for me :)
What is the difference between the 2 `ls` commands that causes the following not to work?
![image](https://github.com/nushell/nushell/assets/24264385/4b79ade9-6f40-43f2-9643-c1aa081f0f65)

**Comment by cronhan** [comment]
I should add that the idea I am working towards is to try to get something like `ls $"($env.USERPROFILE)/*.json"` to work.

**Comment by fdncred** [comment, marked as answer]
You need to use `into glob` in that situation.
```nushell
❯ let toml = "*.toml"
❯ ls $toml <-- fails as expected, similar to your example
Error:   × No matches found for DoNotExpand("*.toml")
   ╭─[entry #7:1:4]
 1 │ ls $toml
   ·    ──┬──
   ·      ╰── Pattern, file or folder not found
   ╰────
  help: no matches found


❯ ls ($toml | into glob)
╭─#─┬────────name─────────┬─type─┬─size──┬───modified───╮
│ 0 │ Cargo.toml          │ file │ 9,391 │ a day ago    │
│ 1 │ Cross.toml          │ file │   684 │ 9 months ago │
│ 2 │ rust-toolchain.toml │ file │ 1,125 │ 3 weeks ago  │
│ 3 │ typos.toml          │ file │   509 │ a day ago    │
╰───┴─────────────────────┴──────┴───────┴──────────────╯
```

**Comment by cronhan** [comment]
Thanks again!

Is this due to the fact that "variables" (aka "constants" as per the docs) are not _actually_ the same as a constant (in this case, "string literal") when it comes to being evaluated?

**Reply by fdncred** [reply]
I don't believe we call variables constants. Please show me.

mut name = val <-- mutable variable
let name = val <-- immutable variable
const name = val <-- constant

We need glob variables to be opt-in because people name things in crazy ways using globbing syntax with *'s, []'s and other illegal characters. If it's opt-in, people can tell nushell that they want globbing.

**Reply by cronhan** [reply]
It is mentioned on this page: https://www.nushell.sh/book/thinking_in_nu.html#variables-are-immutable 
Maybe I misunderstood the intent behind the first paragraph?
As I said, my knowledge and understanding of Nu is not what it should be, and I have lot of "other shell" baggage to check at the door ;)

**Reply by fdncred** [reply]
lol, we all have a lot of other shell baggage, for sure. I think that thinking_in_nu is just a little bit out of date. It was written before there were const's in the language. In a certain perspective it's right because `let` creates immutable variables but in another perspective `let` is not the same as `const`.
